### ISSUE TITLE
Train/Validation Split Bug CNN

### ISSUE THREAD

**Issue by m-c-frank** [BatteryDEV/AutoECM#13]
Hi,
i've been working through your benchmark and have noticed a slight bug in the train/val dataset generation for the CNN part.
[Specifically these lines](https://github.com/BatteryDEV/AutoECM/blob/d893efd2ed6c277d7766dd513459f01e6de1de83/clf_cnn.py#L152-L173):
https://github.com/BatteryDEV/AutoECM/blob/d893efd2ed6c277d7766dd513459f01e6de1de83/clf_cnn.py#L152-L173

The shuffle operation is performed before taking the train/validation cuts so the validation set will almost certainly contain images from the training set (the intersection of the resulting datasets will be approximately as big as the validation_split value).
I also implemented a small test to illustrate that problem (see at the end of this issue).

I think there was just a bit of confusion as to how the seed parameter works.
In order to ensure that the train and validation sets are disjoint you need to set the seed parameter to the same value for both datasets AND set both shuffle flags to the same value. This will ensure that the same random permutation is used for both datasets and thus no intersection will exist.

All in all I don't think this will have a huge impact on the test results of the models but I thought its worth noting because I was confused for a while why my validation accuracy was so high and I think this also impacts the way hyperparameters are tuned/selected. As we could no longer trust the validation accuracy to be a good indicator of the model performance. Especially when using early stopping and other callbacks that rely on the validation accuracy.

I hope this helps anyone else stumbling across.
And I think the simplest fix is to just set the shuffle parameter to True for both datasets? But I know just doing that now would probably interfere with some reproducibility concerns.

Thanks,
Martin Christoph Frank

[The short test script:](https://gist.github.com/m-c-frank/c4d0ddc949fe09a2e67690c6dfef1c29.js)

```python
import tensorflow as tf

INPUT_PATH = "data/images"
SEED = 20
SHUFFLE_TRAIN = True
SHUFFLE_VAL = False
VAL_SPLIT = 0.1


def create_datasets(input_dir):
    train_ds = tf.keras.utils.image_dataset_from_directory(
        f"{input_dir}/train",
        validation_split=VAL_SPLIT,
        subset="training",
        seed=SEED,
        shuffle=SHUFFLE_TRAIN,
    )

    val_ds = tf.keras.utils.image_dataset_from_directory(
        f"{input_dir}/train",
        validation_split=VAL_SPLIT,
        subset="validation",
        seed=SEED,
        batch_size=64,
        shuffle=SHUFFLE_VAL,
    )

    return train_ds, val_ds


if __name__ == "__main__":
    train_ds, val_ds = create_datasets(INPUT_PATH)

    train_paths = train_ds.file_paths
    val_paths = val_ds.file_paths

    # check for intersection
    intersection = set(train_paths).intersection(set(val_paths))
    if len(intersection) > 0:
        print("Intersection exists")
        print(f"n_train: {len(train_paths)}")
        print(f"n_val: {len(val_paths)}")
        print(
            f"n_intersection: {len(intersection)} ({100*len(intersection)/len(train_paths):.2f} %)"
        )
    else:
        print("No intersection")
        print(f"n_train: {len(train_paths)}")
        print(f"n_val: {len(val_paths)}")
        print(f"n_intersection: {len(intersection)}")
```

**Comment by rgasper** [comment]
Ah yeah that's definitely the case. Good catch, Martin.

Raymond Gasper

**Comment by JoachimSchaeffer** [comment]
@m-c-frank Thanks for raising this point. 
I created a new branch to address this issue. 

The CNN model accuracies improve slightly.  This is not a major issue. I also put a note in the ReadMe reading the reproducibility. Please look at my PR and comment in case I missed anything.

Thanks again!

**Comment by JoachimSchaeffer** [comment]
Closing this issue. See commit [bd2e979](https://github.com/BatteryDEV/AutoECM/commit/bd2e9792322c1cb1ccefda6a9e62c39628e43ba5)
Thanks everyone.
